### ISSUE TITLE
map failing to inline leading to nonoptimal codegen

### ISSUE THREAD

**Issue by c42f** [SimonDanisch/FixedSizeArrays.jl#120]
I started looking at the codegen for various ways in which `map` can be used inside elementwise operators.  It seems really hard to get both julia 0.4 and 0.5 happy here, since the optimizer appears to be ignoring `:meta :inline` directives.  See julialang/julia#16712

Some test code I was using to diagnose things:

``` julia
@show FixedSizeArrays.unrolled_map_expr(:func, FixedSizeArrays.SimilarTo{Vec{2,Int}}, size(Vec{2,Int}), (Vec{2,Int},Vector{Int}), (:arg1,:arg2))

@code_native Vec(1,2) + [1,2]

println("\n\n")
immutable Coord <: FixedVectorNoTuple{3,Float64}
    x::Float64
    y::Float64
    z::Float64
end

@show FixedSizeArrays.unrolled_map_expr(:func, FixedSizeArrays.SimilarTo{Coord}, size(Coord), (Coord,Coord), (:arg1,:arg2))

@code_native Coord(1,2,3) + Coord(1,2,3)


println("\n\n")
@show FixedSizeArrays.unrolled_map_expr(:func, FixedSizeArrays.SimilarTo{Vec{2,Int}}, size(Vec{2,Int}), (Vec{2,Int},Vec{2,Int}), (:arg1,:arg2))


@code_native Vec(1,2) + Vec(1,2)
```

The underlying problem is that `map` isn't getting inlined into `+` consistently, and I don't know how to convince the optimizer to do this, I've tried a bunch of different ways of rewriting things.  Some tweaks help for combinations of 0.4, 0.5, FixedVector, FixedVectorNoTuple, Vector etc... but nothing seems to work across the board yet.

**Comment by c42f** [comment]
Looking a bit further, this hasn't really gotten worse in 0.4 due to recent changes, but the generated code is still not ideal.  In 0.5 there's some things which can be tweaked, though 0.5 is such a moving target it's hard to know if they'll stick...

**Comment by SimonDanisch** [comment]
Yeah, I noticed this issue too, but it turned out to be a game of whack-a-mole :(
We should probably open a Julia issue... It sounds pretty buggy, when a function marked as inline doesn't inline ;)

**Comment by c42f** [comment]
Yeah see JuliaLang/julia#16712.

**Comment by SimonDanisch** [comment]
Oh nice, one step ahead :)

**Comment by c42f** [comment]
I think I've now solved this specific issue via JuliaLang/julia#17004 .  There's probably several other codegen issues, but we can tackle them separately.
